### PR TITLE
refactor(www): Create PageMetadata component

### PR DESCRIPTION
**Pull request by tesseralis**
## Description

Create an abstraction over the raw `<Helmet>` component to consolidate common metadata.

TODO:

* [x] Test that everything makes sense and it actually works
* ~~Add more/standardize more descriptions and information~~
  * I'll make another issue for this

## Motivation

* Deduplicate common metadata used between Facebook, Twitter, and other crawlers.
* Deduplicate text in metadata that needs to be translated.

## Build

https://build-5fdd9a37-e2ae-4a55-a681-3bea24c30edb.gtsb.io/

## Related Issues

fixes: #21562

### CODE REVIEW

File: www/src/templates/template-blog-post.js [hunk 1]
```diff
@@ -54,32 +54,16 @@ class BlogPostTemplate extends React.Component {
...
           <main id={`reach-skip-nav`} className="post docSearch-content">
-            {/* Add long list of social meta tags */}
-            <Helmet>
-              <title>{post.frontmatter.title}</title>
+            <PageMetadata
+              title={post.frontmatter.title}
+              description={post.fields.excerpt}
+              image={post.frontmatter.image.childImageSharp.resize}
+            >
+              <meta property="og:url" content={href} />
```
Comment (tesseralis) [inline www/src/templates/template-blog-post.js]: TODO: Fix #21563 in this PR and get rid of this.

File: www/src/templates/template-creator-details.js [hunk 2]
```diff
@@ -68,9 +68,7 @@ class CreatorTemplate extends Component {
 
     return (
       <React.Fragment>
-        <Helmet>
-          <title>{`${creator.name} - Creator`}</title>
-        </Helmet>
+        <PageMetadata title={`${creator.name} - Creator`} />
```
Comment (tesseralis) [inline www/src/templates/template-creator-details.js]: We could possibly add a `description` and `image` to this page if we'd like.

File: www/src/templates/template-starter-page.js [hunk 3]
```diff
@@ -56,33 +56,13 @@ class StarterTemplate extends React.Component {
           maxWidth: 1080,
         }}
       >
+        <PageMetadata
+          title={`${repoName}: Gatsby Starter`}
+          image={screenshot.childImageSharp.fluid}
+          description={`Gatsby Starters: ${repoName}`}
+          timeToRead={1}
```
Comment (tesseralis) [inline www/src/templates/template-starter-page.js]: Are the hardcoded "timeToRead" labels in pages like this useful? We may as well get rid of them.

Comment (tesseralis) [reply www/src/templates/template-starter-page.js]: Actually... I'm trying to find information on `reading time` but I can't seem to find anything. Was this a feature that was taken out at some point?

Comment (tesseralis) [reply www/src/templates/template-starter-page.js]: I've found information on `twitter:label1` only on third-party sites and it seems like it's been removed from the official [twitter developer docs](https://developer.twitter.com/en/docs/tweets/optimize-with-cards/guides/getting-started). Playing around with the [validator](https://cards-dev.twitter.com/validator) doesn't show our time to read in any of our docs.

Comment (muescha) [reply www/src/templates/template-starter-page.js]: for example there is a PR with tweak options for `timeToRead` and some code places where it comes from:

- #19763 `feat(gatsby-transformer-remark, gatsby-plugin-mdx) Make timeToRead configurable`

Comment (LekoArts) [reply www/src/templates/template-starter-page.js]: `twitter:label1` is used in e.g. Slack for richer preview

image

So for Slack `twitter:label1` and `twitter:label2` are valid and were added for this reason.

Comment (LekoArts) [reply www/src/templates/template-starter-page.js]: As for your first question: I wouldn't say they are useful for starters and would switch the information for something like e.g. author or GitHub link.

Comment (tesseralis) [reply www/src/templates/template-starter-page.js]: well that's not intuitive at all. I'm adding a comment :P

Comment (tesseralis) [reply www/src/templates/template-starter-page.js]: I think switching the information is a good idea, but I'll leave that for a separate PR for now.

**Comment by muescha** [comment]
`twitter:image` and `twitter:description` is not anymore in `PageMetadata`?

**Comment by tesseralis** [comment]
@muescha According to the [Twitter documentation](https://developer.twitter.com/en/docs/tweets/optimize-with-cards/guides/getting-started):

> When the Twitter card processor looks for tags on a page, it first checks for the Twitter-specific property, and if not present, falls back to the supported Open Graph property.

So we don't actually need these Twitter descriptions to generate a card (I think). You can verify this behavior by testing the build above with the [Twitter card validator](https://cards-dev.twitter.com/validator).

**Comment by muescha** [comment]
> build-5fdd9a37-e2ae-4a55-a681-3bea24c30edb.gtsb.io

is this still the current build with all new commits?

File: www/src/components/layout.js [hunk 4]
```diff
@@ -20,7 +20,7 @@ export default function DefaultLayout({ location, locale, children }) {
   return (
     <LocaleContext.Provider value={locale || defaultLang}>
       <Global styles={globalStyles} />
-      <SiteMetadata pathname={location.pathname} locale={locale} />
+      <SiteMetadata href={location.href} locale={locale} />
```
Comment (muescha) [inline www/src/components/layout.js]: when the above link is the current status link to the build system, 

then `location.href` is not available at SSR:

Check:
```bash
curl https://build-5fdd9a37-e2ae-4a55-a681-3bea24c30edb.gtsb.io/showcase/reactjs.org | grep "canonical" | less
```
results in:
```html
<link data-react-helmet="true" rel="canonical"/>
<meta data-react-helmet="true" name="docsearch:version" content="2.0"/>
<meta data-react-helmet="true" name="viewport" content="width=device-width,initial-scale=1,shrink-to-fit=no,viewport-fit=cover"/>
<meta data-react-helmet="true" property="og:url"/>
```

while 
```bash
curl https://www.gatsbyjs.org/showcase/reactjs.org | grep "canonical" | less
```

```html
<link data-react-helmet="true" rel="canonical" href="https://www.gatsbyjs.org/showcase/reactjs.org"/>
<meta data-react-helmet="true" name="docsearch:version" content="2.0"/>
<meta data-react-helmet="true" name="viewport" content="width=device-width,initial-scale=1,shrink-to-fit=no,viewport-fit=cover"/>
<meta data-react-helmet="true" property="og:url" content="https://www.gatsbyjs.org"/>
```

Comment (tesseralis) [reply www/src/components/layout.js]: Thanks for the tip! I'll change it back to using pathname. I didn't know that.

Comment (muescha) [reply www/src/components/layout.js]: but check it before with a current build

i remember there was that problem, but unfortunly i can not find some issues or some docs ([troubleshooting](https://www.gatsbyjs.org/docs/troubleshooting-common-errors/)) about this ..

Comment (muescha) [reply www/src/components/layout.js]: added issue:
- `docs: add info about location.href` 

thanks for the link to the blog post in code comments :)

**Comment by AishaBlake** [comment]
Thanks for this (especially the tests), @tesseralis!

**Pull request merged**